tensorflow/core/kernels/dml_batch_to_space_op.cc

```diff
@@ -316,96 +353,165 @@ class DmlBatchToSpaceKernel : public DmlKernel {
-                                                   1, std::multiplies<int64>());
+    dml::Expression output_before_slice;
 
-    // Reshape the input into [block_shape[0], ..., block_shape[M-1], batch /
-    // prod(block_shape), input_shape[1], ..., input_shape[N-1]]
-    dml::TensorDesc::Dimensions reshaped_sizes;
-    for (int i = 0; i < internal_block_sizes.size(); ++i) {
-      reshaped_sizes.push_back(internal_block_sizes[i]);
-    }
+    // For 4d cases we can use the DML Depth to Space operator
+    if (internal_input_shape.dims() == 4) {
+      auto input_sizes = input_desc.desc.GetSizes();
+      auto input_strides = input_desc.desc.GetStrides();
 
-    reshaped_sizes.push_back(internal_input_shape.dim_size(0) /
-                             block_shape_product);
+      if (input_strides.empty())
+      {
+        input_strides = ComputePackedStrides(input_sizes);
+      }
 
-    for (int i = 1; i < internal_input_shape.dims(); ++i) {
-      reshaped_sizes.push_back(internal_input_shape.dim_size(i));
-    }
+      // Permute input NCHW->WNCH
+      dml::TensorDimensions input_tensor_sizes = {
+          input_sizes[3], input_sizes[0], input_sizes[1], input_sizes[2]};
+      dml::TensorStrides input_tensor_strides = {
+          input_strides[3], input_strides[0], input_strides[1],
+          input_strides[2]};
+      input = dml::Reinterpret(input, input_tensor_sizes, input_tensor_strides);
+
+      const auto old_policy = scope.GetTensorPolicy();
+      const auto out_policy = dml::TensorPolicy(
+          [](DML_TENSOR_DATA_TYPE dataType, DML_TENSOR_FLAGS flags,
+             dml::Span<const uint32_t> sizes) {
+            uint32_t dimension_count = static_cast<uint32_t>(sizes.size());
+            // Permute output sizes WNCH->NCHW in order to calculate strides.
+            dml::TensorDimensions output_tensor_sizes = {sizes[1], sizes[2],
+                                                         sizes[3], sizes[0]};
+
+            // Calculate strides with the permuted sizes
+            dml::TensorStrides strides = ComputePackedStrides(output_tensor_sizes);
+
+            // Permute output strides NCHW->WNCH
+            dml::TensorStrides output_tensor_strides = {strides[3], strides[0],
+                                                        strides[1], strides[2]};
+
+            dml::TensorProperties props = {};
+            props.guaranteedBaseOffsetAlignment = 0;
+            props.strides = std::move(output_tensor_strides);
+            props.totalTensorSizeInBytes = DMLCalcBufferTensorSize(
+                dataType, dimension_count, sizes.data(), props.strides->data());
+            return props;
+          });
+
+      // We want to have special output striding for this one operator so it can
+      // mimic BatchToSpace
+      scope.SetTensorPolicy(out_policy);
+      output_before_slice = dml::DepthToSpace(input, internal_block_sizes[0]);
+      scope.SetTensorPolicy(old_policy);
```

Comment:
Couldn't this whole permutation logic be avoided by simply using a layout logic similar to what `dml_conv_ops.cc` does? For example:

```
    // The dimensions of the filter tensor are laid out a little differently
    // than what DML expects
    auto filter_layout = {H, W, C, N};

    // The layout of the input/output tensors is determined by the "data_format"
    // attribute
    auto input_output_layout =
        GetDmlTensorLayout(init_helper->GetDataFormat(), kNchwDimensionCount);
```

Reply:
The naming for Dml::TensorDesc vs DmlTensorDesc is so confusing. 
Internally that layout code seems to be reordering strides like I am. It would 
I like that I can retrieve the Dml::TensorDesc from the expressions because it feels more contained. I think for the input it would be easy but for the output I am not sure if the stride hacking I do is compatible with the logic in that function